### ISSUE TITLE
Provide handling for ASG 'rebalance' events

### ISSUE THREAD

**Issue by spolspol** [aws/aws-node-termination-handler#95]
**Background:**

We run EKS cluster on SPOT instances managed behind an ASG with K8S _cluster-autoscaler_.
This is pretty obvious use case for SPOTs, leveraging Kubernetes self-healing and self-regulating features and elasticity provided by ASG.

With around 9-12 instances, we often get intensive SPOT rotation events, where we loose multiple SPOTs within 10-20 mins - sometimes as much as half the cluster.

When SPOT instances are lost, they are quickly replaced by ASG. However subjected to spare capacity across Availability Zones, instances often come back in the numbers but unbalanced across Zones.

ASG has a rebalancing mechanism and over time it will bring new SPOTs to rebalance ASG across AZs. Rebalanced instances are terminated by ASG from an overflowing AZ.

**Issue:**
While _aws-node-termination-handler_ is handling SPOT terminations well, draining the pods and all, it remains completely oblivious to rebalancing events that ensue.

**Comment by spolspol** [comment]
In above use case scenario we have THREE common termination events.

1) ASG scaling DOWN - node draining handled by _**cluster-autoscaler**_
2) SPOT instance termination - node draining handled by _**aws-node-tyerination-handler**_
3) ASG AZ rebalancing - **not handled**

**Comment by bwagner5** [comment]
This is a good supporting issue for handling of generic shutdown events. I've commented on the referenced issue https://github.com/aws/aws-node-termination-handler/issues/14, but basically I believe the ACPI power button event triggering a cordon by aws-node-termination-handler would solve this issue as well as the generic class of out-of-band terminations without prior notice.

**Comment by DP19** [comment]
Another option is to enabled AZRebalance in the Suspend Processes section of the ASG. This will prevent the ASG from automatically trying to balance the nodes across AZ's. While this doesn't really solve the issue of the termination handling, it will stop it from happening in the first place

**Comment by bwagner5** [comment]
Hey everyone! 👋  We have listened to your request to expand the aws-node-termination-handler to include all EC2 instances and not just Spot.  

We have pushed a branch to the repo [queue-processor]( https://github.com/aws/aws-node-termination-handler/tree/queue-processor) that allows aws-node-termination-handler to run as a deployment in your cluster and respond to EventBridge termination events from ASG, EC2, and Spot. 

Check out the new README on that branch and let us know what you think
https://github.com/aws/aws-node-termination-handler/tree/queue-processor

This new functionality is currently in Alpha, but we would love your feedback!
